Bluetooth: Host: Add req/rsp l2cap validation

HaavardRei opened the pull request:
L2CAP channels will now, along with the ident, store the opcode of the pending request. This commit expands the ident lookup function to also compare received response types to this opcode, and will ignore unsolicited responses.

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -3247,6 +3264,7 @@ int bt_l2cap_chan_disconnect(struct bt_l2cap_chan *chan)
 	LOG_DBG("chan %p scid 0x%04x dcid 0x%04x", chan, le_chan->rx.cid, le_chan->tx.cid);
 
 	le_chan->ident = get_ident();
+	le_chan->pending_req = BT_L2CAP_DISCONN_REQ;
```
Shouldnt't this stuff be done inside the `l2cap_chan_send_req()` function? Otherwise, if something fails (e.g. buffer allocation) or races with this code, then you end up with inconsistent/invalid values inside `le_chan`

HaavardRei replied:
I agree that this would be cleaner, but then we would need to rework functionality like in `bt_l2cap_ecred_chan_reconfigure`:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/l2cap.c#L3095-L3105
(Sets ident for multiple channels, but `l2cap_chan_send_req` is used only for one channel, meaning it would need to know how many channels to set ident/requested opcode for). A similar instance is in `l2cap_ecred_conn_req` (and `bt_l2cap_ecred_chan_reconfigure_explicit`)

alwa-nordic reviewed (changes requested):
Looks good other than my change request below.

HaavardRei commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -139,13 +141,15 @@ static struct bt_l2cap_le_chan *l2cap_chan_alloc_cid(struct bt_conn *conn,
-		if (BT_L2CAP_LE_CHAN(chan)->ident == ident) {
+		if ((BT_L2CAP_LE_CHAN(chan)->ident == ident) &&
+		    ((BT_L2CAP_LE_CHAN(chan)->pending_req == req_opcode) ||
+		     (req_opcode == BT_L2CAP_CMD_REJECT))) {
```
Since the `req_opcode` parameter is of type uint8_t, I think assigning 0x100 will cause it to wrap around to 0? Could use 0xFF instead?

HaavardRei replied:
I see, I updated it now!

HaavardRei committed:
Bluetooth: Host: Add req/rsp l2cap validation

L2CAP channels will now, along with the ident, store the opcode of the
pending request. This commit expands the ident lookup function to also
compare received response types to this opcode, and will ignore
unsolicited responses.

Setting of idents for channels are moved after verification of buffer
allocation for the request to be sent. A TODO is added for improving
this functionality at a later time.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

alwa-nordic reviewed (changes requested):
We must address the Copilot comment about `le_disconn_rsp`.
